feat: Return transaction responses by value (closes #764)

raymax0x opened the pull request:
## Feat: Return Transaction Responses by Value

This pull request refactors the following transaction submission functions to return response structs by value instead of pointers:

- `AddInvokeTransaction`
- `AddDeclareTransaction`
- `AddDeployAccountTransaction`

### Why this change?

- **Performance**: Avoids unnecessary heap allocations for small response structs.
- **Code Quality**: Aligns with Go best practices for handling small, immutable structs by returning them as values.
- **Safety**: Reduces the risk of nil pointer dereferences.

---

### ✅ Changes Made

- Updated function signatures in `rpc/write.go` to return response structs by value (`rpc.AddInvokeTransactionResponse`, etc.).
- Modified the `rpc.RpcProvider` and `account.AccountInterface` interfaces accordingly.
- Updated implementations in `account/transaction.go` to handle value-based responses.
- Regenerated mock files via `go generate ./...` to reflect interface changes.
- Updated relevant tests in both `rpc` and `account` packages to reflect the new signatures.

---

### ✅ Verification

- ✅ All tests passed successfully:

![Test Suite Passed](https://github.com/user-attachments/assets/03741f17-1a3b-4480-9571-d07358ad22a6)

- All existing tests pass successfully after the refactor.
- Full test suite can be run with:

  ```bash
  go test ./...

raymax0x committed:
feat: Return transaction responses by value (issue #764)

Refactor: remove pointer from transaction responses

BREAKING CHANGE: The return types of `AddInvokeTransaction`, `AddDeclareTransaction`, and `AddDeployAccountTransaction` methods in the `RpcProvider` interface and `Account` struct have been changed from pointers to values. This change affects the return type of `BuildAndSendInvokeTxn`, `BuildAndSendDeclareTxn`, `BuildAndEstimateDeployAccountTxn` and `SendTransaction` methods in `Account` struct.

This commit refactors the transaction-related methods in the `rpc` and `account` packages to remove unnecessary pointers from the responses. This change improves code readability and reduces the risk of nil pointer dereferences.

The following changes were made:

- Modified the `AddInvokeTransaction`, `AddDeclareTransaction`, and `AddDeployAccountTransaction` methods in the `RpcProvider` interface and `Provider` struct to return values instead of pointers.
- Updated the `BuildAndSendInvokeTxn`, `BuildAndSendDeclareTxn`, `BuildAndEstimateDeployAccountTxn` and `SendTransaction` methods in the `Account` struct to align with the changes in the `RpcProvider` interface.
- Updated the mock implementations of the `RpcProvider` and `AccountInterface` interfaces to reflect the changes in the method signatures.
- Updated the tests to reflect the changes in the method signatures and return values.

Salonii02 commented on rpc/write.go:
```diff
@@ -11,15 +11,15 @@ import (
 	var output AddInvokeTransactionResponse
 	if err := do(ctx, provider.c, "starknet_addInvokeTransaction", &output, invokeTxn); err != nil {
-		return nil, tryUnwrapToRPCErr(
+		return output, tryUnwrapToRPCErr(
```
I think when there's an error, the function should return an empty struct value, not the potentially partially filled output variable.

raymax0x replied:
Done! Thanks for the feedback Saloni. :)

[commit](https://github.com/NethermindEth/starknet.go/pull/766/commits/4e870486d0628bb906f66f374f5835707ac40ff5) Is this ok ? now.

raymax0x committed:
fix: Return zero-value struct on error.

thiagodeev committed:
Merge branch 'main' into fix/issue-764-return-values

thiagodeev committed:
chore: adjust merged code

thiagodeev commented on account/account.go:
```diff
@@ -28,22 +28,22 @@ type AccountInterface interface {
 		multiplier float64,
 		withQueryBitVersion bool,
-	) (*rpc.BroadcastDeployAccountTxnV3, *felt.Felt, error)
+	) (rpc.BroadcastDeployAccountTxnV3, *felt.Felt, error)
```
This is a full transaction, not a txn response. Let's keep it as a pointer

raymax0x replied:
Understood. Will revert it.

raymax0x replied:
Reverted all the related code for this: [commit](https://github.com/NethermindEth/starknet.go/pull/766/commits/c743d4a545cce7dd96d72647a721421c12a29fc5)

thiagodeev commented on account/transaction.go:
```diff
@@ -28,22 +28,22 @@ import (
 	nonce, err := account.Nonce(ctx)
 	if err != nil {
-		return nil, err
+		return rpc.AddInvokeTransactionResponse{}, err
```
Instead of repeating `rpc.AddInvokeTransactionResponse{}` multiple times, we can create a variable (called `resp` for e.g.) of this type at the top of the method and return it instead. That way, when we have to change anything regarding this in the future, we will only need to edit this variable to reflect the change in the entire method instead of searching for every `rpc.AddInvokeTransactionResponse{}` occurrence and change it.

Do the same for the other `acc.BuildAnd...` methods

raymax0x replied:
Great, that’ll definitely make future tweaks easier. I have refactor the repeated rpc.AddInvokeTransactionResponse{} (and the corresponding acc.BuildAnd… responses) into local resp variables and Made the changes in this commit. Thanks for the suggestion! 

Please refer this commit : [Commit](https://github.com/NethermindEth/starknet.go/pull/766/commits/2b86f52ddad88d52f29338fece3901d97ba2e031)

thiagodeev commented on account/transaction.go:
```diff
@@ -259,57 +259,57 @@ func (account *Account) BuildAndEstimateDeployAccountTxn(
 		resp, err := account.Provider.AddInvokeTransaction(ctx, tx)
 		if err != nil {
-			return nil, err
+			return rpc.TransactionResponse{}, err
```
You can create a top variable of type `rpc.TransactionResponse{}` called `empty` for e.g., and return it instead in all these cases

raymax0x replied:
Made changes in this commit : [Link](https://github.com/NethermindEth/starknet.go/pull/766/commits/2b86f52ddad88d52f29338fece3901d97ba2e031)

thiagodeev reviewed (changes requested):
Thanks @raymax0x!
I left some comments

raymax0x committed:
fix: revert DeployAccount build helper to pointer & sync mocks

• BuildAndEstimateDeployAccountTxn now returns *BroadcastDeployAccountTxnV3
• Adjusted implementation, docs, tests and generated mocks.

All tests pass :
  go generate ./...   # regenerates mocks
  go test ./...       # all packages green

raymax0x committed:
refactor(account): define single zero-value vars to DRY error returns

• BuildAndSendInvokeTxn, BuildAndSendDeclareTxn, DeployContractWithUDC,
  and SendTransaction now declare a `zeroResp` (or equivalent) at the top
  and return it on error instead of repeating literal
  rpc.{AddInvoke,AddDeclare}TransactionResponse{} / TransactionResponse{}.
• No functional change; reduces duplication and eases future type edits.

go test ./...  # still green

raymax0x committed:
docs(changelog): record response-return refactor & zeroResp cleanup

thiagodeev committed:
Merge branch 'main' into fix/issue-764-return-values

thiagodeev committed:
chore: rename zeroResp var, and use it as the return value

thiagodeev committed:
chore: update changelog

thiagodeev reviewed (approved):
Hey @raymax0x! Sorry for the delay.
Thanks!
